### PR TITLE
[github_actions] improve port compatibility

### PR DESCRIPTION
**Pull request by cenit**
Use a parameter to avoid CMake looking for system frameworks on macOS before vcpkg-provided libs.
Reverts #22141 which is unnecessary (if this PR works and is considered effective)

### CODE REVIEW

**Comment by Neumann-A** [comment]
did it find nothing or did it find the wrong file/path?

CMake just does: 
```
find_path(TIFF_INCLUDE_DIR tiff.h)
find_path(PNG_PNG_INCLUDE_DIR png.h PATH_SUFFIXES include/libpng)
```

**Comment by cenit** [comment]
> did it find nothing or did it find the wrong file/path?
> 
> CMake just does:
> 
> ```
> find_path(TIFF_INCLUDE_DIR tiff.h)
> find_path(PNG_PNG_INCLUDE_DIR png.h PATH_SUFFIXES include/libpng)
> ```

it finds wrong headers already installed in system folders (Mono), and on top of that they are incompatible with the ones provided by vcpkg (API breaking changes)

**Comment by cenit** [comment]
yes, it doesn't make sense, i agree. But you can try on a osx github action runner and see by yourself.

**Comment by cenit** [comment]
also because the find_library calls are ok and find vcpkg-built libraries, only the find_path calls are broken and find headers in system folder...

**Comment by cenit** [comment]
adding to the problems, i cannot just remove Mono at the beginning, because I need it to run nuget and setup vcpkg artifact storage 😄

**Comment by Neumann-A** [comment]
> yes, it doesn't make sense, i agree. But you can try on a osx github action runner and see by yourself.
> it's very strange also because the find_library calls are ok and find vcpkg-built libraries, only the find_path calls are broken and find headers in system folder...

I feel like strange is not the right argumentation to merge this PR. Maybe use a overlay until you (or someone) figured out what is really going on. In general, I would expect osx to behave similar so linux so only seeing it on osx is kind of strange too. 

 > it's very strange also because the find_library calls are ok

since `vcpkg.cmake` sets 

```
function(z_vcpkg_add_vcpkg_to_cmake_path list suffix)
    set(vcpkg_paths
        "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}${suffix}"
        "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/debug${suffix}"
    )
    if(NOT DEFINED CMAKE_BUILD_TYPE OR CMAKE_BUILD_TYPE MATCHES "^[Dd][Ee][Bb][Uu][Gg]$")
        list(REVERSE vcpkg_paths) # Debug build: Put Debug paths before Release paths.
    endif()
    if(VCPKG_PREFER_SYSTEM_LIBS)
        list(APPEND "${list}" "${vcpkg_paths}")
    else()
        list(INSERT "${list}" 0 "${vcpkg_paths}") # CMake 3.15 is required for list(PREPEND ...).
    endif()
    set("${list}" "${${list}}" PARENT_SCOPE)
endfunction()
z_vcpkg_add_vcpkg_to_cmake_path(CMAKE_PREFIX_PATH "")
z_vcpkg_add_vcpkg_to_cmake_path(CMAKE_LIBRARY_PATH "/lib/manual-link")
z_vcpkg_add_vcpkg_to_cmake_path(CMAKE_FIND_ROOT_PATH "")
```

it could be that something nuked `CMAKE_PREFIX_PATH` and kept `CMAKE_LIBRARY_PATH` intact

**Comment by dg0yt** [comment]
GDAL added this:
https://github.com/OSGeo/gdal/commit/01968e2eee57d17667d7584018fafe918a149b2c

**Comment by cenit** [comment]
> > yes, it doesn't make sense, i agree. But you can try on a osx github action runner and see by yourself.
> 
> > it's very strange also because the find_library calls are ok and find vcpkg-built libraries, only the find_path calls are broken and find headers in system folder...
> 
> 
> 
> I feel like strange is not the right argumentation to merge this PR. Maybe use a overlay until you (or someone) figured out what is really going on. In general, I would expect osx to behave similar so linux so only seeing it on osx is kind of strange too. 
> 
> 
> 
>  > it's very strange also because the find_library calls are ok
> 
> 
> 
> since `vcpkg.cmake` sets 
> 
> 
> 
> ```
> 
> function(z_vcpkg_add_vcpkg_to_cmake_path list suffix)
> 
>     set(vcpkg_paths
> 
>         "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}${suffix}"
> 
>         "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/debug${suffix}"
> 
>     )
> 
>     if(NOT DEFINED CMAKE_BUILD_TYPE OR CMAKE_BUILD_TYPE MATCHES "^[Dd][Ee][Bb][Uu][Gg]$")
> 
>         list(REVERSE vcpkg_paths) # Debug build: Put Debug paths before Release paths.
> 
>     endif()
> 
>     if(VCPKG_PREFER_SYSTEM_LIBS)
> 
>         list(APPEND "${list}" "${vcpkg_paths}")
> 
>     else()
> 
>         list(INSERT "${list}" 0 "${vcpkg_paths}") # CMake 3.15 is required for list(PREPEND ...).
> 
>     endif()
> 
>     set("${list}" "${${list}}" PARENT_SCOPE)
> 
> endfunction()
> 
> z_vcpkg_add_vcpkg_to_cmake_path(CMAKE_PREFIX_PATH "")
> 
> z_vcpkg_add_vcpkg_to_cmake_path(CMAKE_LIBRARY_PATH "/lib/manual-link")
> 
> z_vcpkg_add_vcpkg_to_cmake_path(CMAKE_FIND_ROOT_PATH "")
> 
> ```
> 
> 
> 
> it could be that something nuked `CMAKE_PREFIX_PATH` and kept `CMAKE_LIBRARY_PATH` intact
> 
> 

i feel confident that under similar setup also windows or linux would show same problem
without full investigation, i'd say that the problem lies in the search order defined inside the module: for example, png.h is under include/libpng16 for us, but the module first searches for png.h in include/libpng, which might roam in system folder before trying with different paths in following lines.
For this reason, it is very difficult to avoid doing this wrapping, which is not dangerous at all, it just makes sure we are choosing vcpkg provided headers.
I am not against this PR in principle, in fact i opened it because i think it's useful and it fixes an open problem for many people, which might not end with these ports. I agree with you that it might be worth a deeper investigation, but why not after this remediation is in place?

**Comment by Neumann-A** [comment]
> I agree with you that it might be worth a deeper investigation, but why not after this remediation is in place?

Because @dg0yt already found the real reason. `vcpkg.cmake` seems to need to set `CMAKE_FIND_APPBUNDLE` and `CMAKE_FIND_FRAMEWORK` to `LAST` instead. In the meantime you can manually pass it to the build. 

>  but the module first searches for png.h in include/libpng

no that is not how it works. 
**PATHS: Specify directories to search in addition to the default locations.** 
in **addition** not _before_ default locations and vcpkg has `png.h` in `/include` so that should be found.

**Comment by dg0yt** [comment]
> i feel confident that under similar setup also windows or linux would show same problem

I disagree. It is about macOS frameworks. See the the link in my other comment.

**Comment by cenit** [comment]
@dg0yt @Neumann-A thanks for the very constructive review
Is it better now?
On my ends, it looks like it's working. Another confirmation would be very welcome, I am on so many different topics right now that i might be missing something.

**Comment by Neumann-A** [comment]
cc @strega-nil-ms since she has more knowledge of osx?

**Review by JackBoosY — APPROVED** [review]

**Review by BillyONeal — APPROVED** [review]
This appears to be in keeping with what vcpkg wants to do (have our bits be found first) based on my reading of the documentation for [`CMAKE_FIND_FRAMEWORK`](https://cmake.org/cmake/help/latest/variable/CMAKE_FIND_FRAMEWORK.html).

We may want to kick off a full rebuild since this touches `buildsystems/vcpkg.cmake` and we found out recently that checking that is probably warranted :)

**Comment by BillyONeal** [comment]
Full rebuild submitted: https://dev.azure.com/vcpkg/public/_build/results?buildId=67909&view=results

**Comment by cenit** [comment]
while this looks enough to fix ga runners, which are nuked from the vcpkg point of view by a mono framework, the discussion about setting also app bundles to last is still open

**Comment by BillyONeal** [comment]
Hmm OK I canceled full rebuild pending that discussion.

**Comment by cenit** [comment]
i feel confident that we are not building frameworks in any of our port. But are we building any app bundle somewhere? we should not, considering that we use a common folder structure unix style, but i might be overlooking something. Shall we put also that condition, to be on the safe side, and then trigger the full rebuild?

**Comment by Neumann-A** [comment]
>  But are we building any app bundle somewhere? 

The cmake test port is an app bundle. Qt and paraview are also building app bundles? (at least they have a folder ending in .app)

x-plane has a .framework folder.

File: scripts/buildsystems/vcpkg.cmake [hunk 1]
```diff
@@ -397,6 +397,7 @@ function(z_vcpkg_add_vcpkg_to_cmake_path list suffix)
         list(APPEND "${list}" "${vcpkg_paths}")
     else()
         list(INSERT "${list}" 0 "${vcpkg_paths}") # CMake 3.15 is required for list(PREPEND ...).
+        set(CMAKE_FIND_FRAMEWORK "LAST" PARENT_SCOPE)
```
Comment (strega-nil-ms) [inline scripts/buildsystems/vcpkg.cmake:400]: This should be on line 407 instead; there's no reason to place this here.

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]

**Comment by strega-nil-ms** [comment]
I am also of the opinion that this is the correct way to do things, since I think that frameworks are unlikely to be built by our ports; we prefer non-framework, unix-style ports. If a port builds a framework, it is also _extremely_ unlikely that the user has a non-framework version of that package on the system.

File: scripts/buildsystems/vcpkg.cmake (hunk 1 above)
Comment (cenit) [reply scripts/buildsystems/vcpkg.cmake:400]: to prefer system frameworks if user wants so, like system libs?

Comment (cenit) [reply scripts/buildsystems/vcpkg.cmake:400]: @strega-nil-ms isn't it reasonable?

**Comment by cenit** [comment]
any news here? I'd like to have github actions fixed on macOS if possible
I still think the line is ok where it is now (ensuring the users can still prefer system-wide frameworks), but waiting on @strega-nil-ms to reply to that question

**Comment by cenit** [comment]
already discussed with @strega-nil-ms on discord
understood the point, will fix the pr asap

**Comment by cenit** [comment]
fixes done as requested

**Comment by cenit** [comment]
@BillyONeal if you want to trigger a full catalog rebuild, this pr is ready

**Comment by JackBoosY** [comment]
@BillyONeal Ready to go now.

**Comment by vicroms** [comment]
It's been a while since I've had to trigger a full rebuild, hope this is correct:

https://dev.azure.com/vcpkg/public/_build/results?buildId=68331&view=results

**Comment by BillyONeal** [comment]
> It's been a while since I've had to trigger a full rebuild, hope this is correct:
> 
> https://dev.azure.com/vcpkg/public/_build/results?buildId=68331&view=results
> 
> Edit: Well... the dreaded msiexec error has appeared

What you did is how I did it before but @ras0219 pointed out that you can queue the build against refs/pull/#/head by typing in the box rather than using the dropdown: 
image

**Comment by cenit** [comment]
x64-osx, the only one that would have been touched by this PR, is green.
The failures in the other triplets might be baseline issues?

**Comment by cenit** [comment]
there is a warning when using this PR:

```
CMake Warning (dev) at F:/vcpkg/scripts/buildsystems/vcpkg.cmake:408 (set):
  Cannot set "CMAKE_FIND_FRAMEWORK": current scope has no parent.
CMake Warning (dev) at F:/vcpkg/scripts/buildsystems/vcpkg.cmake:409 (set):
  Cannot set "CMAKE_FIND_APPBUNDLE": current scope has no parent.
```

**Comment by JackBoosY** [comment]
@cenit Maybe related to the cmake update?

File: scripts/buildsystems/vcpkg.cmake [hunk 2]
```diff
@@ -404,6 +404,11 @@ z_vcpkg_add_vcpkg_to_cmake_path(CMAKE_PREFIX_PATH "")
 z_vcpkg_add_vcpkg_to_cmake_path(CMAKE_LIBRARY_PATH "/lib/manual-link")
 z_vcpkg_add_vcpkg_to_cmake_path(CMAKE_FIND_ROOT_PATH "")
 
+if(NOT VCPKG_PREFER_SYSTEM_LIBS)
+    set(CMAKE_FIND_FRAMEWORK "LAST" PARENT_SCOPE) # we assume that frameworks are usually system-wide libs, not vcpkg-built
+    set(CMAKE_FIND_APPBUNDLE "LAST" PARENT_SCOPE) # we assume that appbundles are usually system-wide libs, not vcpkg-built
```
Comment (dg0yt) [inline scripts/buildsystems/vcpkg.cmake:409]: This is the top-level scope. You cannot use PARENT_SCOPE here.

**Review by dg0yt — CHANGES_REQUESTED** [review]

**Comment by cenit** [comment]
anything more requested here? thie related issue is really blocking here for us, I still hope for a fix (this PR or anything else that might help)

**Comment by Neumann-A** [comment]
> thie related issue is really blocking here for us, I still hope for a fix

It probably will get fixed but it shouldn't be blocking you. As I said you can simply pass the same parameter directly to cmake in the meantime.

**Comment by cenit** [comment]
it would require a tool chain chain loaded and more customizations for github and osx than what is reasonable for a uniform script. Passing the variable to cmake in the downstream project is not propagated upstream in the vcpkg dependencies, if this is what you mean (i think not because you know cmake well)

**Comment by cenit** [comment]
many thanks for merging ❤️

**Comment by fwcd** [comment]
Sorry for necroing this thread... deprioritizing macOS frameworks has some possibly unintended consequences when it comes to system frameworks. As a specific example, if the user has [XQuartz](https://www.xquartz.org/) (a popular X11 server for macOS) installed, CMake will start linking *its* OpenGL implementation at `/usr/X11R6/lib/libGL.dylib` rather than the standard system-provided `OpenGL.framework` (which is rarely the right thing to do).

We ran into this in Mixxx (https://github.com/mixxxdj/mixxx/pull/13080), but it turns out to affect other OpenGL-based apps, causing downstream build breakages in some vcpkg ports that depend on Qt, see #38141

**Comment by cenit** [comment]
is it possible to work on the OpenGL port to promote macOS system framework in that case?

**Pull request merged**